nibabel/tests/test_image_types.py

```diff
@@ -121,3 +123,29 @@ def test_sniff_and_guessed_image_type_randomized():
     img_klasses = copy.copy(all_image_classes)
     np.random.shuffle(img_klasses)
     test_sniff_and_guessed_image_type(img_klasses=img_klasses)
+
+
+def test_sniff_invalid_nii_files():
+    invalid_images = [
+        'invalid.nii',
+        'invalid.nii.gz',
+        'invalid.nii.bz2',
+        #  'invalid.nii.zst',  # requires package pyzstd
+    ]
+
+    invalid_images = [pjoin(DATA_PATH, p) for p in invalid_images]
+
+    for filename in invalid_images:
+        sniff = None
+        for image_klass in all_image_classes:
+            is_valid, sniff = image_klass.path_maybe_image(filename, sniff)
+            assert not is_valid
+            if is_valid:
+                img = image_klass.from_filename(filename, **kwargs)
+                return img
+
+        # reset sniff to None as we are not interested in the header of the decompressed file
+        sniff = None
+        matches, msg = _signature_matches_extension(filename, sniff)
+
+        assert matches == True
```

Comment:
Here we're re-implementing (corrected) `nb.load()` logic. However, this passes because it doesn't trigger the error condition being fixed. Instead, we should make sure that `nb.load()` behaves in the correct way (gives the right output or raises the right error) when given various types of inputs.

We already have:

https://github.com/nipy/nibabel/blob/2838c06389165c9563551d35988df793fb15263d/nibabel/tests/test_loadsave.py#L77-L82

This is getting us the right answer, albeit for the wrong reason because `sniff` is a tuple and not a `bytes`. To detect the actual bug, we need a new test with inputs that are actually zipped but don't contain images. You provide some, but we can just as easily create them on the fly.

```Python
from nibabel.openers import Opener

@pytest.mark.parametrize("extension", [".gz", ".bz2"])
def test_load_good_extension_with_bad_data(tmp_path, extension):
    file_path = tmp_path / f"img.nii{extension}"
    with Opener(file_path, "wb") as fobj:
        fobj.write(b"bad")
    with pytest.raises(ImageFileError, match="Cannot work out file type of .*"):
        load(file_path)
```

WDYT?